File: Array_SpiralOrder.cs
using System;
using System.Collections.Generic;

namespace Algorithms
{
    public class Array_SpiralOrder
    {
        /// Time Complexity : O(M*N) 
        // Space Complexity :O(1) 
        // Did this code successfully run on Leetcode :Yes
        // Any problem you faced while coding this :  No
        public IList<int> SpiralOrder(int[][] matrix)
        {

            List<int> result = new List<int>();
            if (matrix == null || matrix.Length == 0) return result;

            int rows = matrix.Length;
            int columns = matrix[0].Length;
            int top = 0, bottom = rows - 1;
            int left = 0, right = columns - 1;

            while (result.Count < rows * columns)
            {
                //left to right 
                for (int col = left; col <= right; col++)
                {
                    result.Add(matrix[top][col]);
                }
                //top to bottom
                for (int row = top + 1; row <= bottom; row++)
                {
                    result.Add(matrix[row][right]);
                }
                if (top != bottom)
                {
                    //right to left

                    for (int col = right - 1; col >= left; col--)
                    {
                        result.Add(matrix[bottom][col]);
                    }
                }
                if (left != right)
                {
                    //bottom to top

                    for (int row = bottom - 1; row > top; row--)
                    {
                        result.Add(matrix[row][left]);
                    }
                }
                left++;
                right--;
                top++;
                bottom--;

            }
            return result;

        }
    }
}

File: array_spiral_order.py
# Time Complexity : O(n) because we are traversing the matrix once
# Space Complexity : O(1) because we are not using any extra space
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach in three sentences only
"""
In this code we are using the up, down, left, right variables to keep track of the boundaries 
and we continue to append elements in a clockwise fashion with the loops
"""

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        # return matrix.pop(0) + self.spiralOrder([list(i) for i in zip(*matrix)][::-1]) if matrix else []
        
        result = []
        rows, columns = len(matrix), len(matrix[0])
        up = left = 0
        right = columns - 1
        down = rows - 1

        while len(result) < rows * columns:
            for col in range(left, right + 1):
                result.append(matrix[up][col])

            for row in range(up + 1, down + 1):
                result.append(matrix[row][right])

            if up != down:
                for col in range(right - 1, left - 1, -1):
                    result.append(matrix[down][col])

            if left != right:
                for row in range(down - 1, up, -1):
                    result.append(matrix[row][left])

            left += 1
            right -= 1
            up += 1
            down -= 1

        return result